app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +65,138 @@ export class PersistenceManager {
-        this.#dataPersistenceFailing = true;
-        captureException(err);
-      }
-      log.error('error setting state in local store:', err);
-    } finally {
-      this.#isExtensionInitialized = true;
-    }
+
+    await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'exclusive' },
+      async () => {
+        try {
+          // atomically set all the keys
+          await this.#localStore.set({
+            data: state,
+            meta,
+          });
+
+          // back up the vault in IndexedDB
+          const keyringController =
+            state.KeyringController as KeyringControllerState;
+          const vault = keyringController?.vault;
+          if (vault) {
+            await this.#db.set({ vault });
+          } else {
+            // if we don't have a vault, remove the backup from IndexedDB
+            // TODO: should we create a backup for the backup when a vault is
+            // removed?
+            await this.#db.remove(['vault']);
+          }
+
+          if (this.#dataPersistenceFailing) {
+            this.#dataPersistenceFailing = false;
+          }
+        } catch (err) {
+          if (!this.#dataPersistenceFailing) {
+            this.#dataPersistenceFailing = true;
+            captureException(err);
+          }
+          log.error('error setting state in local store:', err);
+        } finally {
+          this.#isExtensionInitialized = true;
+        }
+      },
+    );
   }
 
-  async get() {
-    const result = await this.#localStore.get();
+  /**
+   * Retrieves the current state of the local store. If the store is empty,
+   * it returns undefined. If the store is not open, it throws an error.
+   *
+   * @returns The current state of the local store or null if the store is empty.
+   * @throws Error if the vault is missing and a backup vault is found in IndexedDB.
+   * @throws Error if the local store is not open.
+   */
+  async get(): Promise<MetaMaskStorageStructure | null> {
+    await this.open();
 
-    if (isEmpty(result)) {
-      this.#mostRecentRetrievedState = null;
-      return undefined;
-    }
-    if (!this.#isExtensionInitialized) {
-      this.#mostRecentRetrievedState = result;
-    }
-    return result;
+    return await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'shared' },
+      async () => {
+        const result = await this.#localStore.get();
+
+        const keyringController = result?.data?.KeyringController as
+          | KeyringControllerState
+          | undefined;
+        const hasVault = keyringController?.vault;
+
+        if (!hasVault) {
+          // check if we have a backup vault in IndexedDB
+          // if we do, we need to throw an error so that the user can be
+          // prompted to restore it
+          // if we don't, carry on as if everything is fine (it might be)
+          const [backupVault] = (await this.#db.get(['vault'])) as [
+            string | undefined,
+          ];
```

Comment:
Nit: We could update the return type of `DB.get` to be `Promise<unknown[]>` instead of `Promise<unknown>`, as we _do_ know that it's an array. That lets us get rid of this cast:

```suggestion
          const [backupVault] = await this.#db.get(['vault']);
```

Though the type of `backupVault` is still `unknown` in this case. We could add validation to ensure we throw a different error if it's a non-string, potentially? Of course I wouldn't expect that to happen, but if it did, I'm not sure what would happen.